src/interfaces/wasi/wit_generation.rs

```diff
@@ -0,0 +1,269 @@
+
+    let header_len = u32::to_be_bytes(encrypted_header.encrypted_header_bytes.len() as u32);
+
+    let mut result = header_len.to_vec();
+    result.extend_from_slice(&encrypted_header.encrypted_header_bytes);
+    result.extend_from_slice(&encrypted_block);
+    Ok(result)
+}
+
+#[witgen]
+/// Decrypt ABE-ciphertext (decrypt ABE header + decrypt AES)
+pub fn decrypt(user_decryption_key: String, encrypted_data: Vec<u8>) -> Result<String, String> {
+    let user_key = UserDecryptionKey::from_bytes(
+        &hex::decode(user_decryption_key).map_err(|e| e.to_string())?,
+    )
+    .map_err(|e| e.to_string())?;
+
+    //
+    // Recover header from `encrypted_bytes`
+    let mut header_size_bytes = [0; 4];
+    header_size_bytes.copy_from_slice(&encrypted_data.to_vec()[0..4]);
+    let header_size: usize = u32::from_be_bytes(header_size_bytes) as usize;
+
+    // Split header from encrypted data
+    let header = &encrypted_data[4..(4 + header_size)];
+    let encrypted_block = &encrypted_data[(4 + header_size)..];
+    // let header = &encrypted_data[0..472];
+    // let encrypted_block = &encrypted_data[472..];
+
+    let header_ = decrypt_hybrid_header::<Gpsw<Bls12_381>, Aes256GcmCrypto>(&user_key, header)
+        .map_err(|e| e.to_string())?;
+
+    let cleartext = decrypt_hybrid_block::<Gpsw<Bls12_381>, Aes256GcmCrypto, MAX_CLEAR_TEXT_SIZE>(
+        &header_.symmetric_key,
+        &header_.meta_data.uid,
+        0,
+        encrypted_block,
+    )
+    .map_err(|e| e.to_string())?;
+
+    String::from_utf8(cleartext).map_err(|e| e.to_string())
+}
+
+#[witgen]
+/// Generate a delegate user decryption key for the access policy
+pub fn delegate_user_decryption_key(
+    delegation_key: Vec<u8>,
+    user_decryption_key: String,
+    policy: Vec<u8>,
+    access_policy: Option<String>,
+) -> Result<String, String> {
+    let policy = &serde_json::from_slice(&policy).map_err(|e| e.to_string())?;
+    let access_policy = match access_policy {
+        Some(access_policy) => {
+            AccessPolicy::from_boolean_expression(&access_policy).map_err(|e| e.to_string())?
+        }
+        None => crate::core::policy::AccessPolicy::All,
+    };
+
+    let delegation_key = DelegationKey::from_bytes(&delegation_key).map_err(|e| e.to_string())?;
+    let user_key = UserDecryptionKey::from_bytes(
+        &hex::decode(user_decryption_key).map_err(|e| e.to_string())?,
+    )
+    .map_err(|e| e.to_string())?;
+
+    let engine = Engine::<Gpsw<Bls12_381>>::new();
+    let delegation_key = engine
+        .delegate_user_key(policy, &delegation_key, &user_key, &access_policy)
+        .map_err(|e| e.to_string())?
+        .to_string();
+    Ok(delegation_key)
+}
+
+#[witgen]
+/// Rotating ABE attributes
+pub fn rotate_attributes(policy: Vec<u8>, attributes: Vec<Attribute>) -> Result<Vec<u8>, String> {
+    let policy: &mut crate::core::policy::Policy =
+        &mut serde_json::from_slice(&policy).map_err(|e| e.to_string())?;
```

Comment:
isn't `mut` sufficient?
and later on call explicitly with reference like `serde_json::to_vec(&mut policy).map_err(|e| e.to_string())?;`

Reply:
apparently not...

Reply:
```rust
let mut policy: crate::core::policy::Policy =
        serde_json::from_slice(&policy).map_err(|e| e.to_string())?;
...
let new_policy = serde_json::to_vec(&policy).map_err(|e| e.to_string())?;
```
is working for me

Reply:
indeed!